articles/python-generators/index.md

```diff
@@ -0,0 +1,160 @@
+            yield i
+```
+A generator function returns a generator object which can be iterated to get the values.
+```python
+>>>even_integer_generator(10)
+<generator object even_integer_generator at 0x1036c3200>
+```
+
+```python
+>>>list(even_integer_generator(10))
+[0,2,4,6,8,10]
+```
+3. Using Generator expression.
+```python
+even_integers = (i for i in range(n) if i%2==0)
+```
+Given a list of names to convert each name in the list to uppercase.
+```python
+name_list = ["Adam","Eve","John","Doe","Peter","Paul","Kevin"]
+```
+1. User of functions.
+```python
+def names_to_uppercase(names):
+    uppercase_names = []
+    for name in names:
+        name.upper()
+        uppercase_names.append(name)
+    return uppercase_names
+```
+2. Use of generator functions
+```python
+def names_to_uppercase_generator(names):
+    for name in names:
+        name.upper()
+        yield name
+```
+3. User of generator expressions.
+```python
+uppercase_names = (name.upper() for name in name_list)
+```
+
+**Generating Fibonacci sequence using generator functions.**
+The Fibonacci Sequence is the series of numbers where next number is found by adding up the two numbers before it.
+0, 1, 1, 2, 3, 5, 8, 13, 21, 34, ...
+```python
+def fibonacci_gen():
+    trailing,leading = 0,1
+    while True:
+        yield leading
+        trailing,leading = leading, trailing+leading
+```
+To get the first value of the Fibonacci sequence we generated using generator function above we call the `fibonacci_gen()` function and assign it to variable fib. We then call `next()` on the fib variable, `fib.next()` to get the next number in the sequence.
+```python
+>>>fib = fibonacci_gen()
+>>>fib.next()
+```
+To get the list of the first 10 values for the `fibonacci_gen()` function we call the `fib.next()` inside a for loop with a range of 10
+```python
+for _ in range(10):
+    fib.next()
+```
+
+### Profiling Generator Performance
+```python
+>>> import sys
+# Square root of numbers in a range using list comprehension.
+>>> nums_squared_list = [n * 2 for n in range(10000)]
+>>> sys.getsizeof(nums_squared_list)
+87724
+# Square root of numbers in a range using generator functions.
+>>> nums_squared_generator = (n ** 2 for n in range(10000))
+>>> print(sys.getsizeof(nums_squared_generator))
+125
+```
+List comprehension uses 87724 bytes of memory while generator function uses only 125 bytes of memory. By using generators we save memory as compared to list comprehension where a lot of memory is used.
+
+### Building Generator pipelines
+Generator pipelines allow you to string together code to process large  datasets or streams of data without maxing out your machine’s memory.
+when working with pipelines:-
+- Several pipes can be linked together.
+- Items flow one by one through the entire pipeline.
+- Pipeline functionality can be packaged into a callable function.
+  In order to get the longest name from the text file `names.txt`, We create two generator methods.
```

Comment:
Ensure you include a link to the file `names.txt`. You can either upload it and share the link or mention it in the article itself(if it is a short file).